### PR TITLE
Introduce a network watchdog

### PR DESCRIPTION
**Pull request by posterzh**
**Issue**

- Link: https://github.com/nebraltd/hm-diag/issues/343
- Summary:
To prevent the dropping off the network in hotspots, add network watchdog.

**How**

**Screenshots**

**References**

**Checklist**

- [ ] Tests added
- [ ] Cleaned up commit history (rebase!)
- [ ] Documentation added
- [ ] Thought about variable and method names

### CODE REVIEW

File: hw_diag/utilities/network_watchdog.py
Comment (vpetersson) [inline hw_diag/utilities/network_watchdog.py:10]: This needs to be placed in a persistent storage path.

File: hw_diag/utilities/network_watchdog.py
Comment (vpetersson) [inline hw_diag/utilities/network_watchdog.py:45]: This file will grow indefinitely. We need some kind of `logrotate` style implementation here.

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py]: yes, it is implemented now.
- https://github.com/NebraLtd/helium-miner-software/pull/459/files#diff-29d08d4ce056eee6ab4cba496b7a559f37150deb557009f8d3c4fd9fbd3912beR77
- `RESTART_FILE_NAME = '/var/data/restarts.csv'`

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py]: Make sense. For saving the last restart time and restart counters, I am not going to append them to the file but just write to the file in the volume. For the diagnostics purpose, I would like to configure the logging for the watchdog using `logrotate` in the persistent volume.

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py]: `Keystore` is useful for state tracking. It's done.

File: hw_diag/utilities/balena_supervisor.py
Comment (posterzh) [reply hw_diag/utilities/balena_supervisor.py:64]: There is no action when failed to reboot.

File: hw_diag/app.py
Comment (posterzh) [reply hw_diag/app.py:64]: `ensure_network_connection` sounds better?

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:30]: Typo is fixed now. Thanks

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:69]: It doesn't cover all exceptions. The subclasses of `OSError` is meant to be safe to ignore while trying to connect to `8.8.8.8:53`.

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:126]: It's updated.

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:68]: Good catch. It's updated.

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:118]: This check is not a harm but looks like a redundant. I will remove it.

File: hw_diag/utilities/balena_supervisor.py
Comment (pritamghanghas) [reply hw_diag/utilities/balena_supervisor.py:67]: if we don't plan to do a code fix for the error if it shows up in sentry,  we should log it as warning.

File: hw_diag/app.py
Comment (pritamghanghas) [reply hw_diag/app.py]: ensure_internet_connection() as you check for internet as well.

File: hw_diag/utilities/network_watchdog.py
Comment (pritamghanghas) [reply hw_diag/utilities/network_watchdog.py]: it might be useful to force if it is failing all the time. We have devices stuck in our fleets on updating. @KevinWassermann94 does a reboot recover such devices?

Comment (pritamghanghas) [reply hw_diag/utilities/network_watchdog.py]: @posterzh I still think if we keep a counter for balena_failed_reboot, we can assume that balena is stuck somewhere from which it cannot recover and we can force the reboot. A count of 3 which will be equivalent to 3 hours should be enough.

Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py]: Ok, that would work better.

**Comment by pritamghanghas** [comment]
LGTM except for code smells.

File: hw_diag/utilities/dbus_proxy/network_manager.py
Comment (marvinmarnold) [inline hw_diag/utilities/dbus_proxy/network_manager.py:6]: For consistency with other code, `LOGGER = get_logger(__name__)`

File: hw_diag/utilities/network_watchdog.py
Comment (marvinmarnold) [inline hw_diag/utilities/network_watchdog.py:119]: I'm allergic to deeply nested code. Better if broken into smaller methods.

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py]: Actually this log format stems from `pyhelper` and I simplified it since all watchdog logs come from the single python file. 
However for the consistency, I updated to use the same logging format with `pyhelper`.

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py]: Fair enough. I refactored to avoid the deeply nested codes.

File: hw_diag/utilities/network_watchdog.py
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py]: Simple logging format for watchdog("%(asctime)s - [%(levelname)s]:(%(lineno)d) - %(message)s"):
```
2022-06-06 11:04:45,076 - [INFO]:(64) - Watchdog is started!
2022-06-06 11:04:45,083 - [INFO]:(109) - Running the watchdog...
2022-06-06 11:04:45,090 - [INFO]:(81) - Checking the network connectivity.
2022-06-06 11:04:45,101 - [INFO]:(84) - Network manager state: Connected-Global
2022-06-06 11:04:45,264 - [INFO]:(85) - Internet connectivity: True
2022-06-06 11:04:45,273 - [INFO]:(113) - Network is working.
```

Use `pyhelper` logging format for watchdog:
```
2022-06-06 16:19:44,225 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).__init__ -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(83) - Watchdog is started!
2022-06-06 16:19:44,230 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).ensure_network_connection -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(128) - Running the watchdog...
2022-06-06 16:19:44,234 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).is_connected -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(100) - Checking the network connectivity.
2022-06-06 16:19:44,246 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).is_connected -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(103) - Network manager state: Connected-Global
2022-06-06 16:19:44,417 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).is_connected -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(104) - Internet connectivity: True
2022-06-06 16:19:44,426 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).ensure_network_connection -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(133) - Network is working.
```

@marvinmarnold As the network watchdog log format, what about using the simple logging format so that we can focus on the network status without distracting to the location of the log statement?

Comment (marvinmarnold) [reply hw_diag/utilities/network_watchdog.py]: No strong opinion. The changes look good to me but I'll let @pritamghanghas officially approve the PR when he's ready.

**Comment by posterzh** [comment]
This is the sample watchdog log file. I blocked the `8.8.8.8` at the router for some period of time.
```
root@bee81ffc0a08:/var/watchdog# cat watchdog.log 
2022-06-07 02:44:48,304 - [INFO] - (network_watchdog.py:83) - Watchdog is started!
2022-06-07 02:44:48,306 - [INFO] - (network_watchdog.py:128) - Running the watchdog...
2022-06-07 02:44:48,308 - [INFO] - (network_watchdog.py:100) - Checking the network connectivity.
2022-06-07 02:44:48,316 - [INFO] - (network_watchdog.py:103) - Network manager state: Connected-Global
2022-06-07 02:44:48,452 - [INFO] - (network_watchdog.py:104) - Internet connectivity: True
2022-06-07 02:44:48,457 - [INFO] - (network_watchdog.py:133) - Network is working.
2022-06-07 03:44:48,259 - [INFO] - (network_watchdog.py:128) - Running the watchdog...
2022-06-07 03:44:48,261 - [INFO] - (network_watchdog.py:100) - Checking the network connectivity.
2022-06-07 03:44:48,267 - [INFO] - (network_watchdog.py:103) - Network manager state: Connected-Global
2022-06-07 03:44:48,419 - [INFO] - (network_watchdog.py:104) - Internet connectivity: True
2022-06-07 03:44:48,426 - [INFO] - (network_watchdog.py:133) - Network is working.
2022-06-07 04:44:48,268 - [INFO] - (network_watchdog.py:128) - Running the watchdog...
2022-06-07 04:44:48,270 - [INFO] - (network_watchdog.py:100) - Checking the network connectivity.
2022-06-07 04:44:48,276 - [INFO] - (network_watchdog.py:103) - Network manager state: Connected-Global
2022-06-07 04:44:48,416 - [INFO] - (network_watchdog.py:104) - Internet connectivity: True
2022-06-07 04:44:48,422 - [INFO] - (network_watchdog.py:133) - Network is working.
2022-06-07 05:44:48,265 - [INFO] - (network_watchdog.py:128) - Running the watchdog...
2022-06-07 05:44:48,267 - [INFO] - (network_watchdog.py:100) - Checking the network connectivity.
2022-06-07 05:44:48,272 - [INFO] - (network_watchdog.py:103) - Network manager state: Connected-Global
2022-06-07 05:44:58,284 - [INFO] - (network_watchdog.py:96) - internet not accessible: timed out
2022-06-07 05:44:58,286 - [INFO] - (network_watchdog.py:104) - Internet connectivity: False
2022-06-07 05:44:58,293 - [INFO] - (network_watchdog.py:133) - Network is working.
```

File: hw_diag/utilities/network_watchdog.py
Comment (pritamghanghas) [inline hw_diag/utilities/network_watchdog.py:158]: I think the logic for forcing restart  (after regular restart fails) will not work. save_last_restart() will make sure that we never reach force_reboot = True

Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:150]: Ok, up time looks better than saved last_restart time.

**Comment by pritamghanghas** [comment]
we are still missing jitter from acceptance criterion 1. easy to add, apscheduler supports it. 
Update acceptance criterion number 2 based on the discussion with @vpetersson. Updating 2 might require a new ticket to keep track of the fact that we want to revisit this at some point of time after we have gained some insight on how well it works in the field.

**Review by pritamghanghas — APPROVED** [review]

**Pull request merged**
